safenode/src/domain/storage/spends.rs

```diff
@@ -281,14 +281,194 @@ impl Display for SpendStorage {
     }
 }
 
-/// Still thinking of best location for this.
-/// Wanted to make the DbcAddress take a dbc id actually..
-fn dbc_address(dbc_id: &DbcId) -> DbcAddress {
-    DbcAddress::new(get_dbc_name(dbc_id))
-}
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::domain::dbc_genesis::{create_genesis_dbc, split};
+    use sn_dbc::MainKey;
+    use tempfile::tempdir;
+
+    #[tokio::test]
+    async fn write_and_read_100_spends() {
+        // Test that a range of different spends can be stored and read as expected.
+        let number_of_spends = 100;
+        let mut storage = init_file_store();
+        let key = MainKey::random();
+        let dbc = create_genesis_dbc(&key).expect("Genesis creation to succeed.");
+        let dbcs = split(&dbc, &key, number_of_spends).expect("Split to succeed.");
+        let spends: Vec<_> = dbcs
+            .into_iter()
+            .flat_map(|(dbc, _)| dbc.signed_spends)
+            .collect();
+
+        assert_eq!(spends.len(), number_of_spends);
+
+        for spend in spends {
+            storage
+                .try_add(&spend)
+                .await
+                .expect("Failed to write spend.");
+            let read_spend = storage
+                .get(&dbc_address(spend.dbc_id()))
+                .await
+                .expect("Failed to read spend.");
+
+            assert_eq!(spend.to_bytes(), read_spend.to_bytes());
+        }
+    }
+
+    #[tokio::test]
+    async fn try_add_is_idempotent() {
+        let mut storage = init_file_store();
+        let key = MainKey::random();
+        let src_dbc = create_genesis_dbc(&key).expect("Genesis creation to succeed.");
+
+        let dbc = split(&src_dbc, &key, 1).expect("Split to succeed.");
+        let (dbc, _) = &dbc[0];
+        let spend = dbc.signed_spends.last().expect("Should contain a spend.");
+
+        // Adding the exact same spend is idempotent.
+        storage
+            .try_add(spend)
+            .await
+            .expect("First spend should be added.");
+        storage
```

Comment:
Is this operation idempotent because it doesn't produce an error? Or is there any state we need to verify to make sure something wasn't changed?

Reply:
Adding the exact same spend again and again, doesn't change anything. Because it can only exist or not exist.

Reply:
If we however tried to add a spend with same id as an existing, but something else in it is not the same, then we would have an error.

Depending on what is different, that is just an error returned to the client, or if something that amounts to what would qualify as a double spend attempt, then the spend is killed immediately. It is marked as a double spend attempt.